src/DocumentDB.ChangeFeedProcessor/FeedProcessing/ChangeFeedQueryTimeoutDecorator.cs

```diff
@@ -0,0 +1,49 @@
+﻿// ----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  Licensed under the MIT license.
+// ----------------------------------------------------------------
+
+namespace Microsoft.Azure.Documents.ChangeFeedProcessor.FeedProcessing
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.DataAccess;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.Monitoring;
+    using Microsoft.Azure.Documents.ChangeFeedProcessor.Utils;
+    using Microsoft.Azure.Documents.Client;
+
+    internal class ChangeFeedQueryTimeoutDecorator : IChangeFeedDocumentQuery<Document>
+    {
+        private readonly IChangeFeedDocumentQuery<Document> query;
+        private readonly IHealthMonitor monitor;
+        private readonly TimeSpan timeout;
+
+        public ChangeFeedQueryTimeoutDecorator(IChangeFeedDocumentQuery<Document> query, IHealthMonitor monitor, TimeSpan timeout)
+        {
+            this.query = query;
+            this.monitor = monitor;
+            this.timeout = timeout;
+        }
+
+        public bool HasMoreResults => this.query.HasMoreResults;
+
+        public async Task<IFeedResponse<TResult>> ExecuteNextAsync<TResult>(CancellationToken token = default(CancellationToken))
+        {
+            try
+            {
+                return await this.query.ExecuteNextAsync<TResult>(token).AbortOnTimeout(this.timeout, token).ConfigureAwait(false);
+            }
+            catch (TimeoutException e)
+            {
+                await this.monitor
+                    .InspectAsync(new HealthMonitoringRecord(HealthSeverity.Error, MonitoredOperation.ReadChangeFeed, null, e)).ConfigureAwait(false);
```

Comment:
A bit of problem here... There is no way for health monitor to tell on which partition this happened as we don't pass the lease or PartitionId.

Reply:
what about moving ChangeFeedQueryTimeoutDecorator from PartitionController ctor to PartitionProcessorFactory.Create? Then we would create a CF query inside the factory and pass the lease to the decorator